ffi/wasm/lib/high-level.mjs

```diff
@@ -71,3 +71,77 @@ export class BufferWritable {
     return readString(outStringPtr, outStringLen);
   }
 }
+
+const LocaleRegistry = new FinalizationRegistry(ptr => {
+  icu4x.icu4x_locale_destroy(ptr);
+});
+
+export class Locale {
+  constructor(name) {
+    withEncodedString(name, (encoded, len) => {
+      this.underlying = icu4x.icu4x_locale_create(encoded, len);
+    });
+    LocaleRegistry.register(this, this.underlying);
+  }
+}
+
+const StaticDataProviderRegistry = new FinalizationRegistry(ptr => {
+  const dataProviderView = new Uint32Array(icu4x.memory.buffer, ptr, 2);
+  icu4x.icu4x_data_provider_destroy(dataProviderView[0], dataProviderView[1]); // structs are expanded into multiple params
+  icu4x.icu4x_free(ptr);
+});
+
+const ICU4XCreateDataProviderResult = {
+  parse: rtti.StructType({
+    success: [ 8, rtti.ScalarType.bool ]
+  }),
+  size: 12
+};
+
+export class StaticDataProvider {
+  constructor() {
+    const receiveBuffer = icu4x.icu4x_alloc(ICU4XCreateDataProviderResult.size);
+    icu4x.icu4x_static_data_provider_create(receiveBuffer);
+    const resultParsed = ICU4XCreateDataProviderResult.parse(icu4x.memory.buffer, receiveBuffer);
+    if (resultParsed.success) {
+      this.underlying = receiveBuffer; // pointer to where the actual data provider is
+      StaticDataProviderRegistry.register(this, this.underlying);
+    } else {
+      throw new Error("Failed to create a static data provider");
+    }
+  }
+}
+
+const FixedDecimalFormatRegistry = new FinalizationRegistry(ptr => {
+  icu4x.icu4x_fixed_decimal_format_destroy(ptr);
+});
+
+const ICU4XCreateFixedDecimalFormatResult = {
+  parse: rtti.StructType({
+    fdf: [ 0, rtti.ScalarType.u32 ],
+    success: [ 4, rtti.ScalarType.bool ]
+  }),
+  size: 8
+};
+
+export class FixedDecimalFormat {
+  constructor(locale, provider, options) {
+    const receiveBuffer = icu4x.icu4x_alloc(ICU4XCreateFixedDecimalFormatResult.size);
+    icu4x.icu4x_fixed_decimal_format_create(receiveBuffer, locale.underlying, provider.underlying, 0, 0);
+    const resultParsed = ICU4XCreateFixedDecimalFormatResult.parse(icu4x.memory.buffer, receiveBuffer);
+    icu4x.icu4x_free(receiveBuffer);
+
+    if (resultParsed.success) {
+      this.underlying = resultParsed.fdf;
+      FixedDecimalFormatRegistry.register(this, this.underlying);
+    } else {
+      throw new Error("Failed to create a fixed decimal format");
+    }
+  }
+
+  write(decimal, writable) {
```

Comment:
suggestion: should we just directly use the BufferWriteable stuff so that users can directly get a string? both here and in FixedDecimal

(I'd also then call this method `format()`)

Reply:
I'm a bit worried about introducing high-level logic beyond what's available in the Rust API (esp from the perspective of figuring out how diplomat would generate an equivalent API). Should we just special-case APIs that take in a writable?

Reply:
Oh, CustomWriteable is absolutely going to be a diplomat special case for "writing to a string". So yeah. I've done something similar on the C++ side.

Reply:
I ended up having two APIs, one that uses writable and one that uses it under the hood so that users can manually use writables if they need to.